src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurStandardAccountingExtractCollectorBatchBuilder.java

```diff
@@ -0,0 +1,304 @@
+package edu.cornell.kfs.concur.batch.service.impl;
+
+import java.util.Date;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.apache.commons.lang.StringUtils;
+import org.kuali.kfs.gl.batch.CollectorBatch;
+import org.kuali.kfs.gl.businessobject.OriginEntryFull;
+import org.kuali.kfs.sys.KFSConstants;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+
+import edu.cornell.kfs.concur.ConcurConstants;
+import edu.cornell.kfs.concur.ConcurParameterConstants;
+import edu.cornell.kfs.concur.batch.businessobject.ConcurStandardAccountingExtractDetailLine;
+import edu.cornell.kfs.concur.batch.businessobject.ConcurStandardAccountingExtractFile;
+import edu.cornell.kfs.gl.CuGeneralLedgerConstants;
+
+/**
+ * Helper class for converting parsed SAE file data
+ * into GL CollectorBatch objects, which in turn will contain
+ * OriginEntryFull objects derived from the SAE detail lines.
+ * The constructed CollectorBatch objects will have the "HD"
+ * (Header) record type.
+ * 
+ * This builder is NOT thread-safe; external synchronization
+ * is required if one instance is used by multiple threads.
+ */
+public class ConcurStandardAccountingExtractCollectorBatchBuilder {
+
+    private static final org.apache.log4j.Logger LOG = org.apache.log4j.Logger.getLogger(ConcurStandardAccountingExtractCollectorBatchBuilder.class);
+
+    protected static final String FAKE_OBJECT_CODE_PREFIX = "?NONE?";
+    protected static final int INITIAL_BUILDER_SIZE = 50;
+    protected static final int REPORT_ID_LENGTH_FOR_DOC_NUMBER = 10;
+    protected static final int NAME_LENGTH_FOR_DESCRIPTION = 14;
+    protected static final int MIN_BATCH_SEQUENCE_NUMBER = 0;
+    protected static final int MAX_BATCH_SEQUENCE_NUMBER = 9;
+
+    protected final String docTypeCode;
+    protected final String systemOriginationCode;
+    protected final String chartCode;
+    protected final String highestLevelOrgCode;
+    protected final String departmentName;
+    protected final String campusCode;
+    protected final String campusAddress;
+    protected final String notificationEmail;
+    protected final String notificationPerson;
+    protected final String notificationPhone;
+
+    protected CollectorBatch collectorBatch;
+    protected Map<String,OriginEntryFull> originEntries;
+    protected Function<Date,Integer> fiscalYearFinder;
+    protected BiFunction<Date,String,String> dateFormatter;

```

Comment:
I haven't seen this before, would this be a good item for a book club discussion?  It's a cool method